src/sources/docker.rs

```diff
@@ -0,0 +1,888 @@
+
+        let mut message = String::from_utf8(message.data)
+            .map_err(|error| error!(message="Recieved corrupted message",%error))
+            .ok()?;
+
+        let mut splitter = message.splitn(2, char::is_whitespace);
+        let timestamp_str = splitter.next()?;
+        match DateTime::parse_from_rfc3339(timestamp_str) {
+            Ok(timestamp) => {
+                // Timestamp check
+                match self.last_log.as_ref() {
+                    // Recieved log has not already been processed
+                    Some(&(ref last, gen))
+                        if *last < timestamp || (*last == timestamp && gen == self.generation) =>
+                    {
+                        ()
+                    }
+                    // Recieved log is not from before of creation
+                    None if self.created <= timestamp.timestamp() => (),
+                    _ => {
+                        trace!(
+                            message = "Recieved older log",
+                            timestamp = field::display(timestamp_str)
+                        );
+                        return None;
+                    }
+                }
+                // Supply timestamp
+                log_event.insert_explicit(
+                    event::TIMESTAMP.clone(),
+                    timestamp.with_timezone(&chrono::Utc).into(),
+                );
+
+                self.last_log = Some((timestamp, self.generation));
+
+                let log = splitter.next()?;
+                let remove_len = message.len() - log.len();
+                message.drain(..remove_len);
+            }
+            Err(error) => {
+                // Recieved bad timestamp, if any at all.
+                error!(message="Didn't recieve rfc3339 timestamp from docker",%error);
+                // So log whole message
+            }
+        }
+
+        // Message is actually one line from stderr or stdout, and they are delimited with newline
+        // so that newline needs to be removed.
+        message.pop().filter(|&c| c != '\n').map(|c| {
+            // For some reason last character isn't a newline, so return it
+            message.push(c);
+        });
+
+        // Supply message
+        log_event.insert_explicit(event::MESSAGE.clone(), message.into());
+
+        // Supply container
+        log_event.insert_implicit(event::CONTAINER.clone(), self.id.as_str().into());
+
+        let event = Event::Log(log_event);
+        trace!(message = "Received one event", event = field::debug(&event));
+        Some(event)
+    }
+}
+
+#[cfg(all(test, feature = "docker-integration-tests"))]
+mod tests {
+    use super::*;
+    use crate::test_util::{collect_n, trace_init};
+
+    /// None if docker is not present on the system
+    fn source<'a, L: Into<Option<&'a str>>>(
+        name: &str,
+        label: L,
+    ) -> (mpsc::Receiver<Event>, tokio::runtime::Runtime) {
+        let mut rt = tokio::runtime::Runtime::new().unwrap();
```

Comment:
this should use `test_util::runtime()` instead of creating a runtime directly like this.

Reply:
Added